```
$ i586-pc-msdosdjgpp-gcc fpex.c 
fpex.c: In function 'main':
fpex.c:15:13: error: implicit declaration of function '_controlfp'; did you mean '_control87'? [-Wimplicit-function-declaration]
   15 |     fp_cw = _controlfp( fp_cw,
      |             ^~~~~~~~~~
      |             _control87
fpex.c:31:29: error: 'EM_PRECISION' undeclared (first use in this function)
   31 |             status[ (bits & EM_PRECISION) == 0 ] );
      |                             ^~~~~~~~~~~~
fpex.c:31:29: note: each undeclared identifier is reported only once for each function it appears in
```

--- a/.github/workflows/test-suite-worker.yml
+++ b/.github/workflows/test-suite-worker.yml
@@ -51,6 +51,11 @@ jobs:
       id: test_prereq
       run: ./ci_test_prereq.sh
 
+    - name: Open Watcom install
+      uses: open-watcom/setup-watcom@v0
+      with:
+        version: 2.0-64
+
     - name: Test
       id: test
       env:

--- a/test/common_framework.py
+++ b/test/common_framework.py
@@ -392,6 +392,26 @@ class BaseTestCase(object):
         ofile = basename.with_suffix('.com')
         check_call(["nasm", "-f", "bin", "-o", str(ofile), str(sfile)])
 
+    def mkexe_with_watcom(self, fname, content, dname=None, extraargs=None):
+        if dname is None:
+            p = self.workdir
+        else:
+            p = Path(dname).resolve()
+        basename = p / fname
+
+        sfile = basename.with_suffix('.c')
+        sfile.write_text(content)
+
+        watcom = environ.get("WATCOM", '0')
+        if watcom == '0':
+            raise ValueError('WATCOM variable not set')
+        environ['INCLUDE'] = f'{watcom}/h'
+
+        args = ["wcl", "-zq", "-bt=dos", "-fpi87", f"-fe={basename}.exe", f"-fm={basename}.map", sfile]
+        if extraargs:
+            args += extraargs
+        check_call(args)
+
     def mkfile(self, fname, content, dname=None, mode="w", newline=None):
         if dname is None:
             p = self.workdir / fname

--- a/test/func_fpu_bart_exceptions.py
+++ b/test/func_fpu_bart_exceptions.py
@@ -1,0 +1,81 @@
+
+def fpu_bart_exceptions_fpex(self):
+    if self.use_cpu == "kvm":
+        if not self.have_kvm:
+            self.skipTest("requires KVM")
+        cpu_vm = 'kvm'
+    elif self.use_cpu == "emu":
+        cpu_vm = 'emulated'
+    else:
+        raise ValueError('invalid self.use_cpu')
+
+    config = """\
+$_hdimage = "dXXXXs/c:hdtype1 +1"
+$_floppy_a = ""
+$_cpu_vm = "%s"
+""" % cpu_vm
+
+    self.mkfile("testit.bat", """\
+c:\\fpex
+rem end
+""", newline="\r\n")
+
+    # compile sources
+    self.mkexe_with_watcom("fpex", r"""
+#include <stdio.h>
+#include <float.h>
+#include <dos.h>
+
+char *status[2] = { "disabled", "enabled" };
+
+void main()
+{
+    unsigned int fp_cw = 0;
+    unsigned int fp_mask = 0x3f;
+    unsigned int bits;
+    double a = 1.0;
+
+    fp_cw = _controlfp( fp_cw,
+                         fp_mask );
+
+    printf( "Interrupt Exception Masks\n" );
+    bits = fp_cw & MCW_EM;
+    printf( "  Invalid Operation exception %s\n",
+            status[ (bits & EM_INVALID) == 0 ] );
+    printf( "  Denormalized exception %s\n",
+            status[ (bits & EM_DENORMAL) == 0 ] );
+    printf( "  Divide-By-Zero exception %s\n",
+            status[ (bits & EM_ZERODIVIDE) == 0 ] );
+    printf( "  Overflow exception %s\n",
+            status[ (bits & EM_OVERFLOW) == 0 ] );
+    printf( "  Underflow exception %s\n",
+            status[ (bits & EM_UNDERFLOW) == 0 ] );
+    printf( "  Precision exception %s\n",
+            status[ (bits & EM_PRECISION) == 0 ] );
+
+    printf( "Infinity Control = " );
+    bits = fp_cw & MCW_IC;
+    if( bits == IC_AFFINE )     printf( "affine\n" );
+    if( bits == IC_PROJECTIVE ) printf( "projective\n" );
+
+    printf( "Rounding Control = " );
+    bits = fp_cw & MCW_RC;
+    if( bits == RC_NEAR )       printf( "near\n" );
+    if( bits == RC_DOWN )       printf( "down\n" );
+    if( bits == RC_UP )         printf( "up\n" );
+    if( bits == RC_CHOP )       printf( "chop\n" );
+
+    printf( "Precision Control = " );
+    bits = fp_cw & MCW_PC;
+    if( bits == PC_24 )         printf( "24 bits\n" );
+    if( bits == PC_53 )         printf( "53 bits\n" );
+    if( bits == PC_64 )         printf( "64 bits\n" );
+    printf("%g\n", a/0.0);
+    printf("FAIL:\n");
+}
+""")
+
+    results = self.runDosemu("testit.bat", config=config)
+
+    self.assertIn('Floating point exception', results)
+    self.assertNotIn('FAIL:', results)

--- a/test/func_fpu_bart_exceptions.py
+++ b/test/func_fpu_bart_exceptions.py
@@ -1,3 +1,5 @@
+from common_framework import DOSEMU_CONF_DEFAULT
+
 
 def fpu_bart_exceptions_fpex(self):
     if self.use_cpu == "kvm":
@@ -9,11 +11,8 @@ def fpu_bart_exceptions_fpex(self):
     else:
         raise ValueError('invalid self.use_cpu')
 
-    config = """\
-$_hdimage = "dXXXXs/c:hdtype1 +1"
-$_floppy_a = ""
-$_cpu_vm = "%s"
-""" % cpu_vm
+    config = DOSEMU_CONF_DEFAULT
+    config += f"""$_cpu_vm = "{cpu_vm}"\n"""
 
     self.mkfile("testit.bat", """\
 c:\\fpex
@@ -78,4 +77,80 @@ void main()
     results = self.runDosemu("testit.bat", config=config)
 
     self.assertIn('Floating point exception', results)
+    self.assertNotIn('FAIL:', results)
+
+
+def fpu_bart_exceptions_fpexes(self):
+    if self.use_cpu == "kvm":
+        if not self.have_kvm:
+            self.skipTest("requires KVM")
+        cpu_vm = 'kvm'
+    elif self.use_cpu == "emu":
+        cpu_vm = 'emulated'
+    else:
+        raise ValueError('invalid self.use_cpu')
+
+    config = DOSEMU_CONF_DEFAULT
+    config += f"""$_cpu_vm = "{cpu_vm}"\n"""
+
+    self.mkfile("testit.bat", """\
+c:\\fpexes
+rem end
+""", newline="\r\n")
+
+    self.mkcom_with_nasm("fpexes", r"""
+org 100h
+    mov ax, 3502h
+    int 21h
+    push es
+    push bx
+    mov dx, int2_handler
+    mov ax, 2502h
+    int 21h
+    fninit
+    fnstcw [cw]
+    and word [cw], ~0x3f
+    fldcw [cw]
+    fld1
+    fldz
+    fdiv
+    fwait
+    pop dx
+    pop ds
+    mov ax, 2502h
+    int 21h
+    fninit
+    ret
+
+int2_handler:
+    sti
+    push ax
+    push dx
+    push ds
+    fstsw ax
+    fnclex
+    test ax, 80h
+    mov ah, 9
+    push cs
+    pop ds
+    mov dx, esclearstr
+    jz esclear
+    mov dx, essetstr
+esclear:
+    int 21h
+    pop ds
+    pop dx
+    pop ax
+    iret
+
+essetstr:
+    db 'PASS:ES set',13,10,'$'
+esclearstr:
+    db 'FAIL:ES clear',13,10,'$'
+cw:
+""")
+
+    results = self.runDosemu("testit.bat", config=config)
+
+    self.assertIn('PASS:', results)
     self.assertNotIn('FAIL:', results)

--- a/test/test_processor.py
+++ b/test/test_processor.py
@@ -11,7 +11,7 @@ from func_build_freedos import build_freedos
 from func_build_pcmos import build_pcmos
 from func_cpu_trap_flag import cpu_trap_flag
 from func_cpu_methods import cpu_create_items
-from func_fpu_bart_exceptions import fpu_bart_exceptions_fpex
+from func_fpu_bart_exceptions import fpu_bart_exceptions_fpex, fpu_bart_exceptions_fpexes
 from fpu.qemu import fpu_create_items
 
 
@@ -41,6 +41,11 @@ class OurTestCase(BaseTestCase):
     def test_fpu_bart_exceptions_fpex(self):
         """FPU Exceptions (Bart) (fpex)"""
         fpu_bart_exceptions_fpex(self)
+
+    @mark('fputest')
+    def test_fpu_bart_exceptions_fpexes(self):
+        """FPU Exceptions (Bart) (fpexes)"""
+        fpu_bart_exceptions_fpexes(self)
 
 
 class KVMTestCase(ppdosgit(OurTestCase, {

--- a/test/test_processor.py
+++ b/test/test_processor.py
@@ -11,6 +11,7 @@ from func_build_freedos import build_freedos
 from func_build_pcmos import build_pcmos
 from func_cpu_trap_flag import cpu_trap_flag
 from func_cpu_methods import cpu_create_items
+from func_fpu_bart_exceptions import fpu_bart_exceptions_fpex
 from fpu.qemu import fpu_create_items
 
 
@@ -35,6 +36,11 @@ class OurTestCase(BaseTestCase):
     def test_cpu_trap_flag(self):
         """CPU Trap Flag"""
         cpu_trap_flag(self)
+
+    @mark('fputest')
+    def test_fpu_bart_exceptions_fpex(self):
+        """FPU Exceptions (Bart) (fpex)"""
+        fpu_bart_exceptions_fpex(self)
 
 
 class KVMTestCase(ppdosgit(OurTestCase, {

